### PR TITLE
Add CMake Project Template

### PR DESCRIPTION
**Pull request by ProtoByter**
This PR contains a CMake Script that I wrote for my game engine (which uses CMake) which pratically converts what is done in the ppu_rules makefile into CMake

I've tested this successfully with the VideoTest sample (both running via the .self file and via the .pkg file)

I hope this is useful :)

### CODE REVIEW

**Comment by bengl3rt** [comment]
Awesome!

**Comment by bengl3rt** [comment]
I ended up adding another macro, setupPSL1GHTLibrary, that doesn't do any of the elf file manipulation stuff at the end; that allowed some of my code to be built as a static library (.a file) for linking by a latter (S)ELF.

`
macro(setupPSL1GHTLibrary target)

    if(NOT PSL1GHT)

        message(FATAL_ERROR "Location of PSL1GHT SDK not set")

    endif()

    set(CMAKE_C_COMPILER ${PSL1GHT}/ppu/bin/powerpc64-ps3-elf-gcc)

    set(CMAKE_CXX_COMPILER ${PSL1GHT}/ppu/bin/powerpc64-ps3-elf-g++)

    target_link_directories(${target} PUBLIC ${PSL1GHT}/ppu/lib ${PSL1GHT}/spu/lib ${PSL1GHT}/portlibs/ppu/lib)

    target_include_directories(${target} PUBLIC ${PSL1GHT}/ppu/include ${PSL1GHT}/spu/spu/include ${PSL1GHT}/portlibs/ppu/include)

    target_compile_options(${target} PUBLIC -mcpu=cell)

    target_link_libraries(${target} PUBLIC gcm_sys simdmath rsx sysutil io lv2 rt audio m)

endmacro()

`

**Comment by ProtoByter** [comment]
I think I might've added that functionality locally but I probably forgot to commit it

**Comment by miigotu** [comment]
I was going to comment before that last revert 🤣

**Comment by ProtoByter** [comment]
Oh shit, sorry
I forgot about this PR and I've recently had issues with the Makefile using `?=` and not `:=` when setting `LD` so I just changed it on my fork

**Comment by ProtoByter** [comment]
Should I close this?

**Comment by miigotu** [comment]
> Should I close this?

No I think if it leaves the original functionality in ace and adds cmake as an alternative build mechanism it is a fine addition. I'll have to look over it and check your conditionals and other changes when I'm at a PC.

**Comment by ProtoByter** [comment]
> No I think if it leaves the original functionality in ace.

The bigger issue is that currently it doesn't, it's replaced the `?=` in the `base_rules` file with a `:=` because I was having issues with it trying to use my system `ld` which gave me errors about the `.o` file being in being

**Comment by ProtoByter** [comment]
Wrong button :/

**Comment by ProtoByter** [comment]
Anyway if I were to do it now I'd probably make a lot of changes since right now it uses a lot of weirdness (ie the fact that `CMAKE_C_COMPILER` is set in the function)
I need to look into how `nxdk` does it since they recently switched to cmake and from what I've seen it looks really good

**Comment by miigotu** [comment]
> > No I think if it leaves the original functionality in ace.
> 
> The bigger issue is that currently it doesn't, it's replaced the `?=` in the `base_rules` file with a `:=` because I was having issues with it trying to use my system `ld` which gave me errors about the `.o` file being in the wrong format

I'm not sure if that is what the issue is/was? I don't see how immediate vs deferred variable assignment would change which ld was used, unless your cmake sets it incorrectly somehow. Maybe we can figure out a different way to handle that issue.

**Comment by ProtoByter** [comment]
> > > No I think if it leaves the original functionality in ace.
> > 
> > 
> > The bigger issue is that currently it doesn't, it's replaced the `?=` in the `base_rules` file with a `:=` because I was having issues with it trying to use my system `ld` which gave me errors about the `.o` file being in the wrong format
> 
> I'm not sure if that is what the issue is/was? I don't see how immediate vs deferred variable assignment would change which ld was used, unless your cmake sets it incorrectly somehow. Maybe we can figure out a different way to handle that issue.
 
Oh, I wasn't using CMake for this, I was using the 'trivial' template project. I was just saying that, that's a change I made unrelated to CMake, so I would need to close the PR and make a new branch and do it from there.

**Comment by miigotu** [comment]
You probably need an spu.cmake and ppu.cmake, which sets the prefix and adds it to the linker arguments?

For ppu `PREFIX=ppu-` and for spu `PREFIX=spu-`, they are set in spu_rules and ppu_rules in the project root. I don't think you included those anywhere.

```cmake
set(CMAKE_CXX_LINK_EXECUTABLE "${PREFIX} ${CMAKE_CXX_LINK_EXECUTABLE}")
```

File: templates/cmake/cmake/psl1ght.cmake [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+endif()
+
+add_custom_target(PS3Package)
+
+macro(setupPSL1GHT target)
+    if(NOT PSL1GHT)
+        message(FATAL_ERROR "Location of PSL1GHT SDK not set")
+    endif()
+    set(CMAKE_C_COMPILER ${PSL1GHT}/ppu/bin/powerpc64-ps3-elf-gcc)
+    set(CMAKE_CXX_COMPILER ${PSL1GHT}/ppu/bin/powerpc64-ps3-elf-g++)
```
Comment (miigotu) [inline templates/cmake/cmake/psl1ght.cmake:12]: Add these after line 12:
```cmake
set(CMAKE_CXX_LINKER ${PSL1GHT}/ppu/bin/powerpc64-ps3-elf-gcc)
set(CMAKE_C_LINKER ${PSL1GHT}/ppu/bin/powerpc64-ps3-elf-gcc)
```

I think setting prefix as a variable at the top of the file cleans up your code a bit.

File: base_rules [hunk 2]
```diff
@@ -21,7 +21,7 @@ export AS		:=	$(PREFIX)as
 export CC		:=	$(PREFIX)gcc
 export CXX		:=	$(PREFIX)g++
 export AR		:=	$(PREFIX)ar
-export LD		?=	$(PREFIX)gcc
+export LD		:=	$(PREFIX)gcc
```
Comment (miigotu) [inline base_rules:24]: Change this back

**Comment by ProtoByter** [comment]
So, I'm going to close this and make a CMake build system for PSL1GHT, once I've finished, I'll open a new PR
Does that sound good?

**Comment by miigotu** [comment]
Sure does, so far we have had unity in psl1ght build systems, but maybe a new build system as an option will stir up some new contributions.

**Comment by ProtoByter** [comment]
Ok, closing

**Pull request closed without merging**